test-data/unit/check-unreachable-code.test

```diff
@@ -457,3 +458,44 @@ else:
 reveal_type(x)  # E: Revealed type is 'builtins.str'
 [builtins fixtures/ops.pyi]
 [out]
+
+[case testShortCircuitInExpression]
+import typing
+def f() -> bool: pass
+PY2 = f()
+PY3 = f()
+
+a = PY2 and 's'
+b = PY3 and 's'
+c = PY2 or 's'
+d = PY3 or 's'
+reveal_type(a)  # E: Revealed type is 'builtins.bool'
```

Comment:
Oh wait this is testing the special-casing for PY2 and PY3 as well. I guess it's okay. I may not get to the full review until next week.